### ISSUE TITLE
Conditional relationships

### ISSUE THREAD

**Issue by zepfietje** [AdditionApps/flexible-presenter#5]
It would be really nice if relations could be added to the presenter based on if they're already loaded on the model or not (just like https://laravel.com/docs/7.x/eloquent-resources#conditional-relationships).

This issue is just for discussing the behavior. Possibly, I'll create a PR soon.

**Comment by fourstacks** [comment]
Hi @zepfietje, could you give me an idea of the syntax that you'd want here?  For example do you want to use a presenter with a nested relation but only when that relation is loaded.  E.g:

```
// in PostsPresenter
public function values(): array
    {
        return [
            'id' => $this->id,
            'title' => $this->title,
            'comments' => CommentsPresenter::make($this->whenLoaded('comments')),
        ];
    }
```

It's worth pointing out that you can sort of pull this off to a degree using the `lazy` method e.g.

```
'comments' => $this->lazy(CommentsPresenter::make($this->comments)),
```

This will only evaluate the `comments` key when it's actually requested (meaning that you can choose not to load the relation and you won't get any errors about the relation not existing if you do not call for that key).  That does still rely on you (as the developer) being aware of which relationships you've loaded when calling presenters though so I'll admit it introduces more cognitive load than the approach in Laravel Resources.

I like the idea though realistically it'll be about another month before I can put time towards this so feel free to submit a PR if you end up whipping up something.

**Comment by zepfietje** [comment]
Hi @fourstacks,
The syntax I envisioned is exactly as you showed in your first example and as available in Laravel Resources. I already knew about the `lazy` method, but indeed think it would be a cool addition to be able to handle the conditional relationships in the presenter instead of having to specify relations using the `only` and `except` methods.

We'd still have to figure out what's the best way for conditional relationships to work in conjunction with `only` and `except` (which overrules which?).

I'll try to implement it when I can find the time!

**Comment by fourstacks** [comment]
Hey @zepfietje - just to let you know that there's now a `whenLoaded` method in v1.2.1.  It works broadly the same way as it does in API resources just without the additional parameters that the framework method accepts (the presenter version just takes the relationship name).

It's tested but let me know if you run into any problems with it :)

**Comment by zepfietje** [comment]
@fourstacks, thanks! Looks good. Only thing to note is that when the relation is not loaded and you're using `whenLoaded`, the relation key is included but the value is `null`. When using `only`, the key is omitted entirely. I know this behavior is as designed and probably not such a big problem, but still wanted to mention it. Thanks again for implementing this so quickly!

**Comment by fourstacks** [comment]
@zepfietje no problem!

Regarding your comment about relation loading I'm not sure that's quite the behaviour in eloquent models.

If you do not load the relation (either using `Post::with('comments')->find(1)` or `$post->load('comments')` then the relation will not appear in the `relations` property of the model.  

E.g. if I were to `dd($post)` I'd see this amongst the keys:

```
...
  #relations: []
...
``` 
However if I load the relation but there are no related models I'd see this if it's a to-one type of relation:

```
...
  #relations: [
    'author' => null
  ]
...
``` 

Or this for a one to many type of relation:

```
...
  #relations: [
    'comments' => Illuminate\Database\Eloquent\Collection] {
      #items: []
    }
  ]
...
``` 

However that has thrown up an interesting issue - I'd be interested to get your take on it seeing as you're also wanting to use this method:

So if dealing with a to-one type of relation:

`'author' => UserPresenter::make($this->whenLoaded('author'))`

This will yield `null` if the `author` relation is not loaded OR if it is loaded but there is no related model

Both return points are consistent so that's all good.  However consider this (a to-many relation type):

`'comments' => CommentPresenter::collection($this->whenLoaded('comments'))`

This will yield `null` if the `comments` relation is not loaded BUT if it will currently yield an empty array if it is loaded but there are no related comments.

My question is whether the return type should always be null if either the relationship is not loaded or it is loaded but no related models exist.  Would you prefer to deal with null or an empty array?

**Comment by zepfietje** [comment]
@fourstacks, interesting point.

I believe it makes sense to return `null` and `[]` when a relation is loaded but empty for a to-one and to-many relationship respectively. So the current behavior is perfect.

However, for the `whenLoaded` part I think it might be neat to omit the relation key entirely when it is not loaded, just like when using `only()`. This is also how Laravel Resources handle it, and according to your comment, Eloquent models go about it similarly.

In short, the current behavior regarding empty relations is perfect. But when `whenLoaded` is used and a relation is not loaded, the relation key should be omitted from the presenter entirely.

**Comment by fourstacks** [comment]
Ahh - gotcha!  Sorry, I misunderstood your previous message I think - apologies for the unnecessary explanation on the `relations` key within models - I can see now that's not what you were referring to.

OK, so  I think you're right - it makes sense to stick to the same convention used in API Resources where the key is omitted if the relation is not loaded.  I'll drop that change in asap and push up a new version.

Thanks for kicking this feature around with me - helps to have someone else sense check these things!

**Comment by zepfietje** [comment]
@fourstacks, no problem.

I'm more than happy to help. Let me know when you've implemented the change!

**Comment by fourstacks** [comment]
@zepfietje Just an update on this.  I've had a go at implementing the behaviour we outlined above but it's proving quite tricky to pull off - all my efforts so far have felt quite heavy handed (not to mention that none of them have actually worked across all contexts!).

At this point, this change would also technically be a breaking one so I think overall I might park this idea for now and perhaps revisit as and when a v2 makes sense.  If the feature is really important to you I'll happy take a look at a PR in the meantime though.

Thanks again for your help!

**Comment by zepfietje** [comment]
@fourstacks could you share your attempt? Might help me when I'll give it a shot.

Thanks for your efforts!

**Comment by fourstacks** [comment]
Afraid I nuked the work I'd done as it really wasn't heading anywhere I liked.  The general crux of the idea though was that if `whenLoaded` is called and the relation isn't loaded, the resource is set to a 'marker' class of some sort (rather than null) to make it clear that we're dealing with an unloaded relation (I created an empty class called MissingValue for instance).  I couldn't really work out how to stop this from blowing things up in the `get` method though and I was changing more and more code in an attempt to get things to work which just felt wrong for such a small feature/change.

Hope that's of some (limited) help!
